### PR TITLE
Return mapping optimization in tensor_mechanics

### PR DESCRIPTION
**Pull request by WilkAndy**
Rachel and Andy work - long time coming, but it's finally OK.  Thanks @waxmanr ,

a

### CODE REVIEW

**Comment by permcody** [comment]
@dschwen - You should probably review

File: modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticJ2.h [hunk 1]
```diff
@@ -89,8 +89,95 @@ class TensorMechanicsPlasticJ2 : public TensorMechanicsPlasticModel
...
+    * @param[out] returned_intnl The value of the internal parameter after returning
+    * @param[out] dpm   plastic multipliers needed to ring about the return
+    * @param[out] delta_dp The change in plastic strain induced by the return process
+    * @param[out] The value of the yield function, evaluated at (trial_stress, intnl_old) in the case of return value = false (failure), or at (returned_stress, returned_intnl) for return value = true (success)
+    * @param[out] trial_stress_inadmissible Should be set to 0 if the trial_stress is admissible, and 1 if the trial_stress is inadmissible.  This can be used by the calling prorgram
+    * @return true if a successful return (or a return-map not needed), false if the trial_stress is inadmissible but the return process failed
+    */
+  bool returnMap(const RankTwoTensor & trial_stress, const Real & intnl_old, const RankFourTensor & E_ijkl, Real ep_plastic_tolerance,
+                        RankTwoTensor & returned_stress, Real & returned_intnl, std::vector<Real> & dpm, RankTwoTensor & delta_dp,
+                        std::vector<Real> & yf, unsigned & trial_stress_inadmissible) const;
```
Comment (dschwen) [inline modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticJ2.h]: Any reason `trial_stress_inadmissible` is not a `bool`?

File: modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticModel.h [hunk 2]
```diff
@@ -138,7 +138,8 @@ class TensorMechanicsPlasticModel : public GeneralUserObject
    * @param[out] act act[i] = true if the i_th yield function is active
    * @param[out] returned_stress Approximate value of the returned stress
    */
-  virtual void activeConstraints(const std::vector<Real> & f, const RankTwoTensor & stress, const Real & intnl, const RankFourTensor & Eijkl, std::vector<bool> & act, RankTwoTensor & returned_stress) const;
+  virtual void activeConstraints(const std::vector<Real> & f, const RankTwoTensor & stress, const Real & intnl,
+                                  const RankFourTensor & Eijkl, std::vector<bool> & act, RankTwoTensor & returned_stress) const;
 
   /// Returns the model name (eg "MohrCoulom")
```
Comment (dschwen) [inline modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticModel.h]: typo (MohrCoulomb)

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C [hunk 3]
```diff
@@ -125,8 +128,8 @@ ComputeMultiPlasticityStress::computeQpStress()
   if (_fspb_debug == 3)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: What is `3`? Could an enum be clearer here?

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C [hunk 4]
```diff
@@ -137,14 +140,20 @@ ComputeMultiPlasticityStress::computeQpStress()
...
-  bool found_solution = elasticStep(_stress_old[_qp], _stress[_qp], _intnl_old[_qp], _intnl[_qp], _plastic_strain_old[_qp], _plastic_strain[_qp], _my_elasticity_tensor, _my_strain_increment, _yf[_qp], number_iterations, _Jacobian_mult[_qp]);
+  // plastic multipliers and cumulative plastic multipliers.
+  // Here they are used by quickStep below and are unimportant dummies
+  std::vector<Real> pm;
+  pm.assign(_num_surfaces, 0);
+  std::vector<Real> cumulative_pm;
+  cumulative_pm.assign(_num_surfaces, 0);
+
+  // try a "quick" return first - this can be purely elastic, or a customised plastic return defined by a TensorMechanicsPlasticXXXX UserObject
+  bool found_solution = quickStep(_stress_old[_qp], _stress[_qp], _intnl_old[_qp], _intnl[_qp], pm, cumulative_pm, _plastic_strain_old[_qp], _plastic_strain[_qp], _my_elasticity_tensor, _my_strain_increment, _yf[_qp], number_iterations, _Jacobian_mult[_qp], "computeQpStress", true);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: `found_solution` should be `const` here (the quickStep call might as well be in the if condition)

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C [hunk 5]
```diff
@@ -137,14 +140,20 @@ ComputeMultiPlasticityStress::computeQpStress()
   bool ld_encountered = false;
   bool constraints_added = false;
 
-  // try a purely elastic step first
-  bool found_solution = elasticStep(_stress_old[_qp], _stress[_qp], _intnl_old[_qp], _intnl[_qp], _plastic_strain_old[_qp], _plastic_strain[_qp], _my_elasticity_tensor, _my_strain_increment, _yf[_qp], number_iterations, _Jacobian_mult[_qp]);
+  // plastic multipliers and cumulative plastic multipliers.
+  // Here they are used by quickStep below and are unimportant dummies
+  std::vector<Real> pm;
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: This adds the slow local `std::vector` paradigm back in. I thought we tried to get rid of this.

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C [hunk 6]
```diff
@@ -205,15 +218,70 @@ ComputeMultiPlasticityStress::postReturnMap()
...
+  }
+  else if (num_plastic_returns == 1 && successful_return)
+  {
+    // one model has successfully completed its custom returnMap algorithm
+    // and the other models have signalled they are elastic at
+    // the trial stress
+    plastic_strain = plastic_strain_old + delta_dp;
+    if (final_step)
+    {
+      if (called_from.compare("computeQpStress") == 0)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: A string comparison is likely to be a significant slow down here. I suggest using an enum if you must have a construct like this.

File: modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticJ2.h (hunk 1 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticJ2.h]: changed trial_stress_inadmissible to a bool

File: modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticModel.h (hunk 2 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/include/userobjects/TensorMechanicsPlasticModel.h]: fixed MohrCoulom typo

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C (hunk 4 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: i'm not sure what you mean here, @dschwen .  Do you think the code should read

if (!quickStep(blah,blah))
   plasticStep(blah, blah)

?

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C (hunk 5 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: Are you suggesting making a new global variable in the constructor?

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C (hunk 6 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: used an enum

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C (hunk 4 above)
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: No, the additional bool variable helps the readability, but you should make it `const`. It is just a placeholder that is not intended to be changed.

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C (hunk 5 above)
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: A member variable, yes. One that is only sized at construction time, and not repeatedly allocated and deallocated.

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C (hunk 4 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: aha, i get it.  made found_solution const.  nice one, i hadn't thought of doing that, so you can tell i haven't received any formal training in C.  i wonder how many millions of other places i could do this in my code!

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C (hunk 5 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: made _dummy_pm, which is a member variable.  Also _cumulative_pm member variable, which avoids one extra "assign" in some cases

File: modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C (hunk 3 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/src/materials/ComputeMultiPlasticityStress.C]: yea, enum is better, thanks - that's quite old code that i hadn't looked at in months

**Comment by WilkAndy** [comment]
Thanks @dschwen and @jwpeterson ,

a

**Comment by WilkAndy** [comment]
Hey @waxmanr - not sure if you're still around in the world of MOOSE - but thanks for your work.  Our big PR finally got through :-)

a

**Comment by waxmanr** [comment]
@WilkAndy --- I haven't been using it lately, but I've been following the issues and PRs related to all of this. Thanks for wrapping it up and getting everything merged.

**Pull request merged**